File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.cpp
#include <string>
#include <iostream>
#include <vector>

using namespace std;
int limit[30];

vector<int> solution(string today, vector<string> terms, vector<string> privacies) {
    vector<int> answer;
    
    int Y = stoi(today.substr(0,4));
    int M = stoi(today.substr(5,2));
    int D = stoi(today.substr(8,2));
    
    for (auto t : terms)
    {
        string lim = t.substr(2);
        limit[t[0] - 'A'] = stoi(lim);
    }
    
    int idx = 0;
    for (auto pri : privacies)
    {
        idx++;
        int y = stoi(pri.substr(0,4));
        int m = stoi(pri.substr(5,2));
        int d = stoi(pri.substr(8,2));
        char kind = pri[11];
        
        int deleteM = m + limit[kind - 'A'];
        y += limit[kind - 'A'] / 12;
        m += limit[kind - 'A'] - 12 * (limit[kind - 'A'] / 12);
        if (m >= 13) y++, m %= 12;
        
        cout << kind << " : " << y << " " << m << " " << d << '\n';
        
        if (y > Y) continue;
        if (y == Y && m > M) continue;
        if (y == Y && m == M && d > D) continue;
        
        answer.emplace_back(idx);
    }
    
    return answer;
}

File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.java
import java.io.*;
import java.util.*;

class Solution {
    
    Map<String,Integer> M = new HashMap<>();
    
    public String calc(String date, String kind) {
        StringTokenizer st = new StringTokenizer(date, ".");
        int year = Integer.parseInt(st.nextToken());
        int month = Integer.parseInt(st.nextToken());
        int period = M.get(kind);
        
        month += period;
        while (month > 12) {
            year++;
            month -= 12;
        }
        
        String m = (month < 10) ? "0" + Integer.toString(month) : Integer.toString(month);
        return year + "." + m + "." + st.nextToken();
    }
    
    public int[] solution(String today, String[] terms, String[] privacies) {
        int n = terms.length;
        
        for (int i = 0; i < n; i++) {
            String term = terms[i];
            StringTokenizer st = new StringTokenizer(term);
            String a = st.nextToken();
            int b = Integer.parseInt(st.nextToken());
            M.put(a, b);
        }
        
        
        
        List<Integer> endList = new ArrayList<>();
        for (int i = 0; i < privacies.length; i++) {
            StringTokenizer st = new StringTokenizer(privacies[i]);
            String a = st.nextToken();
            String b = st.nextToken();
            String deadLine = calc(a, b);
            
            // System.out.println(deadLine);
            if (today.compareTo(deadLine) >= 0) endList.add(i + 1);
            
        }
        
        
        int[] answer = new int[endList.size()];
        for (int i = 0; i < endList.size(); i++) {
            answer[i] = endList.get(i);
        }
        return answer;
    }
}